build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1670,228 @@ message WaitExecutionRequest {
+message GetPotentialActionResultsResponse {
+  // A set of possible matches to the client filesystem along with a reference to build outputs.
+  repeated ObservedPathSetAndInputDigest potential_action_matches = 1;
+
+  // If present, signifies that there are more results which the client can
+  // retrieve by passing this as the page_token in a subsequent
+  // [request][build.bazel.remote.execution.v2.GetPotentialActionResults].
+  // If empty, signifies that this is the last page of results.
+  string next_page_token = 2;
+}
+
+// The ordered file reads and probes and directory enumerations used in
+// two-phase caching to compare to the client filesystem when seeking a match
+// from an action fingerprint formed from potentially incomplete
+// (underspecified) inputs to a set of outputs.
+//
+// When comparing to the filesystem, a corresponding observed-input digest is
+// initialized. Each pathset entry is examined against the filesystem, and
+// depending on the outcome of the comparison, new hash information is added
+// to the observed-input digest. At the end of pathset enumeration the
+// observed-input digest is compared to the observed-input digest provided in
+// the two-phase cache lookup (`observed_input_digest` in
+// [ObservedPathSetAndInputDigest][build.bazel.remote.execution.v2.ObservedPathSetAndInputDigest])
+// and if they match, this is a cache hit, and the resulting
+// [ActionResult][build.bazel.remote.execution.v2.ActionResult] should be
+// retrieved and used.
+//
+// Generating a directory hash: Create a list of the names of directory members
+// (files and directories) that pass the directory enumeration or
+// `ObservedAccessFileNames` filters. Lexicographically sort the names by code
+// point or, equivalently, by the UTF-8 bytes. Join the sorted strings together
+// with the ',' character to form a comma-delimited list; note for a
+// multi-member list this means that there is no tail ',', e.g. for
+// a.h and b.h the list string is "a.h,b.h". Convert this string to uppercase.
+// Hash the UTF-8 string bytes.
+//
+// Discussion of how to generate pathsets from action execution filesystem
+// observations can be found at
+// https://github.com/microsoft/BuildXL/blob/master/Documentation/Wiki/Advanced-Features/Search-Path-Enumeration.md .
+message ObservedPathSet
+{
+  // An individual file or directory access observed during execution of
+  // an [Action][build.bazel.remote.execution.v2.Action].
+  message Entry
+  {
+    // Operation kind.
+    enum OpKind
+    {
+      // `Path` refers to a file that was read by the action. If the file is
+      // not present, this pathset does not match the filesystem. Otherwise
+      // the hash of the string constant "R" followed by the hash of the file
+      // must be added to the observed-input digest.
+      FILE_READ = 0;
+
+      // `Path` is a directory that was enumerated by the action using
+      // `DirectorySearchPattern`. If the directory is missing, this pathset
```

Comment:
Why does DIRECTORY_ENUMERATION use DirectorySearchPattern, but DIRECTORY_SEARCH not?

Reply:
DIR_SEARCH is not a regular enumeration, it's a combination of one or more specific filenames (i.e. no wildcards involved). This is typical for C++ compilers, which (in most implementations) call the filesystem enumeration APIs with exactly the filename of the header/include desired. For example, if a .cpp contains:

#include "foo.h"
#include "bar.h"

And the search paths include "dir1;dir2" the filesystem will typically see the following sequence of directory enumerations:

dir1: enum "foo.h" (it's not here)
dir2: enum "foo.h" (it's here)
dir1: enum "bar.h" (it's here so file probing stops)

In the ObservedPathSet this becomes:
Entries:
    { path="dir1", kind=DIR_SEARCH }
    { path="dir2", kind=DIR_SEARCH }
observed_accessed_file_names = [ "foo", "bar" ]

It's important in these cases to ensure that the presence or absence of foo(.h) and bar(.h) in both dir1 and dir2 are added to the observed input digest as that affects correctness of the cache hits in the face of such directory searching/probing.

Reply:
Is this a Windows-specific API? Why wouldn't the accessed file names be "foo" and "bar"? Is some equivalent of `stat` not used to query if a directory contains a file with a specific name?

Reply:
stat() equivalence on Windows is inexact and apps do it differently. If you have an open file handle there's GetFileInformationByHandle() but the most common case when querying directory metadata is to start a dir enumeration with FindFirstFile() and following FindNextFile(), but using an exact name string like "foo.h" instead of using wildcard characters.

On Linux the sandbox would see a stat() call for a specific name but the resulting ObservedPathSet result is the same: A file search for a specific name, and tracking whether the file was present or not in the directory, to feed the summary hash to use for comparison with what the worker saw from its sandboxed API calls.

Reply:
@alercah I was avoiding adding such an example into the doc comment text but I don't see a great way to add in more information. I did add a link to the BuildXL discussion markdown for reference; do you think that URL would be usefully repeated on the per-field comments as well to aid understanding?

Reply:
Ah I see from the comment above that you found that doc impenetrable. Let me create a better one or modify the existing and see if that works better.